dev-tools/validation/test_design_capability_allocation.py

```
@@ -0,0 +1,256 @@
+def fixture_root() -> tempfile.TemporaryDirectory[str]:
+    fixture = tempfile.TemporaryDirectory()
+    shutil.copytree(ROOT / "design", Path(fixture.name) / "design")
+    return fixture
+
+
+def expect_failure(label: str, validator, root: Path, expected: str) -> None:
+    try:
+        validator(root)
+    except SystemExit as error:
+        if expected not in str(error):
+            raise AssertionError(f"{label}: unexpected failure: {error}") from error
```

Ruff (0.15.21)

[warning] Avoid specifying long messages outside the exception class

(TRY003)